src/scripts.rs

```diff
@@ -38,6 +44,109 @@ use serde::de;
 /// <https://github.com/bitcoin/bitcoin/blob/4a540683ec40393d6369da1a9e02e45614db936d/src/primitives/transaction.h#L87-L89>
 pub const SEQUENCE_LOCKTIME_MASK: u32 = 0x00_00_ff_ff;
 
+/// A public key used in derived descriptors
+#[derive(Debug, Eq, PartialEq, Clone, Ord, PartialOrd, Hash)]
+pub struct DerivedPublicKey {
+    /// Fingerprint of the master xpub and the derivation index used. We don't use a path
+    /// since we never derive at more than one depth.
+    pub origin: (bip32::Fingerprint, bip32::ChildNumber),
+    /// The actual key
+    pub key: PublicKey,
+}
+
+impl fmt::Display for DerivedPublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let (fingerprint, deriv_index) = &self.origin;
+
+        write!(f, "[")?;
+        for byte in fingerprint.as_bytes().iter() {
+            write!(f, "{:02x}", byte)?;
+        }
+        write!(f, "/{}", deriv_index)?;
+        write!(f, "]{}", self.key)
+    }
+}
+
+impl FromStr for DerivedPublicKey {
+    type Err = ScriptCreationError;
+
+    fn from_str(s: &str) -> Result<DerivedPublicKey, Self::Err> {
+        // The key is always of the form:
+        // [ fingerprint / index ]<key>
+
+        // 1 + 8 + 1 + 1 + 1 + 66 minimum
+        if s.len() < 78 {
+            return Err(ScriptCreationError::DerivedKeyParsing);
+        }
+
+        // Non-ASCII?
+        for ch in s.as_bytes() {
+            if *ch < 20 || *ch > 127 {
+                return Err(ScriptCreationError::DerivedKeyParsing);
+            }
+        }
+
+        if s.chars().next().expect("Size checked above") != '[' {
+            return Err(ScriptCreationError::DerivedKeyParsing);
+        }
+
+        let mut parts = s[1..].split(']');
+        let fg_deriv = parts
+            .next()
+            .ok_or_else(|| ScriptCreationError::DerivedKeyParsing)?;
```

Comment:
Oh i didn't `clippy` @danielabrozzoni will tell me to replace by `ok_or` here :sweat_smile: